provider/malicious_sources/resource_email_domain_alert.go

```diff
@@ -0,0 +1,194 @@
+package malicious_sources
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"strings"
+	"github.com/traceableai/terraform-provider-traceable/provider/common"
+	"github.com/traceableai/terraform-provider-traceable/provider/custom_signature"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func ResourceEmailDomainAlert() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceEmailDomainAlertCreate,
+		Read:   resourceEmailDomainAlertRead,
+		Update: resourceEmailDomainAlertUpdate,
+		Delete: resourceEmailDomainAlertDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:        schema.TypeString,
+				Description: "name of the email domain policy",
+				Required:    true,
+			},
+			"description": {
+				Type:        schema.TypeString,
+				Description: "description of the policy",
+				Optional:    true,
+			},
+			"rule_action": {
+				Type:        schema.TypeString,
+				Description: "Need to provide the action to be performed ",
+				Optional:    true,
+				Default: "BLOCK",
+			},
+			"event_severity": {
+				Type:        schema.TypeString,
+				Description: "Generated event severity among LOW,MEDIUM,HIGH,CRITICAL",
+				Required:    true,
+			},
+			"environment": {
+				Type:        schema.TypeSet,
+				Description: "environment where it will be applied",
+				Required:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"data_leaked_email": {
+				Type:        schema.TypeBool,
+				Description: "Users from leaked email domain",
+				Required:    true,
+			},
+			"disposable_email_domain": {
+				Type:        schema.TypeBool,
+				Description: "Users from disposable email domain",
+				Required:    true,
+			},
+			"email_domains": {
+				Type:        schema.TypeSet,
+				Description: "Email domains for rule",
+				Required:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"email_regexes": {
+				Type:        schema.TypeSet,
+				Description: "Email regexes for rule",
+				Required:    true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+			},
+			"email_fraud_score": {
+				Type:        schema.TypeString,
+				Description: "Minimum email fraud score",
+				Required:    true,
+			},
+		},
+	}
+}
+
+func resourceEmailDomainAlertCreate(d *schema.ResourceData, meta interface{}) error {
+	event_severity := d.Get("event_severity").(string)
+	name := d.Get("name").(string)
+	rule_action := d.Get("rule_action").(string)
```

Comment:
There is lot of repetition of code in create update and read while building a query.. intstead you can take the query build part in a helper fucntion and call it in each method

Reply:
For each type there are different input sets so i would anyways required to create 5 to 6 different helpers and then return a map of these values and then read from it one by one